### DISCUSSION TITLE
Is there any way to dynamically inject pcode?

### DISCUSSION THREAD

**Discussion by Escapingbug** [NationalSecurityAgency/ghidra#5450] [Q&A]
I need a way to modify P-code. According to my findings, the P-code is generated by SLEIGH each time (for decompilation).
A `PcodeInjectLibrary` could possibly change the Pcode on-the-fly which could further influence the decompilation result.
But somehow I only found that `PcodeInjectLibrary` is defined by compiler spec statically, which means for a single architecture this class is already defined and not dynamically modifiable.
Do we have any other way to achieve dynamically modifiable `PcodeInjectLibrary`?
This could be really helpful as many analyses can be achieved by modifying the pcode without patching the binary since the binary patching is restricted.

**Comment by Danil6969** [comment]
Maybe write a script to write context registers? You shouldn't directly inject anything because of incapsulation principle anyway as it may be unsafe. Just provide an interface to do any reconstruction work you need in narrow place(s) only. It will be like:
> scan other context registers, save and don't modify them

> clear code, only edit desired context register, disassemble again

> match another pattern and not the first one later on

> see changes in the decompiler.

I'll post example with exceptions injection and it's corresponding script to make it clear

**Comment by Danil6969** [comment]
See these two, ia.sinc (named ia.txt, important lines are 327-328 and 1790-1799) and SetExcept.java (named SetExcept.txt)
[ia.txt](https://github.com/NationalSecurityAgency/ghidra/files/11725557/ia.txt)
[SetExcept.txt](https://github.com/NationalSecurityAgency/ghidra/files/11725580/SetExcept.txt)
This is an edit required to impelement "except" as block which is usually found in pascal/delphi programs but it may be substituted with "try", "catch", "finally" (this is found in java mostly) etc. The decompiler should detect these keywords and print them as special blocks by omitting "if" and its braces. "if" is required to block constant folding and alike rules. But overall it's correct anyway since executor "enters" the block only if exception was thrown and flag is raised and skips the block otherwise.

**Comment by Escapingbug** [comment]
> See these two, ia.sinc (named ia.txt, important lines are 327-328 and 1790-1799) and SetExcept.java (named SetExcept.txt) [ia.txt](https://github.com/NationalSecurityAgency/ghidra/files/11725557/ia.txt) [SetExcept.txt](https://github.com/NationalSecurityAgency/ghidra/files/11725580/SetExcept.txt) This is an edit required to impelement "except" as block which is usually found in pascal/delphi programs but it may be substituted with "try", "catch", "finally" (this is found in java mostly) etc. The decompiler should detect these keywords and print them as special blocks by omitting "if" and its braces. "if" is required to block constant folding and alike rules. But overall it's correct anyway since executor "enters" the block only if exception was thrown and flag is raised and skips the block otherwise.

If I get it correct, it means we will have to modify the *.sinc to add the custom context registers we need. Then how do we load that dynamically? AFAIK, .sinc file is loaded by ghidra statically and selected when selecting architecture, can we modify that when program already loaded?

**Comment by Danil6969** [comment]
It's pretty easy. First marshal all the data which requires interface already done through e.g. context registers. Then scan their value through sinc/slaspec. You only will be unable to randomly inject pcode at runtime which is absent in sinc/slaspec. But it doesn't mean that you can't transfer data to disassembler so it changes pattern selection itself.

**Comment by Escapingbug** [comment]
> It's pretty easy. First marshal all the data which requires interface already done through e.g. context registers. Then scan their value through sinc/slaspec. You only will be unable to randomly inject pcode at runtime which is absent in sinc/slaspec. But it doesn't mean that you can't transfer data to disassembler so it changes pattern selection itself.

Forgive me, but I don't quite understand what you mean.
My question is I need to modify the semantic of certain pieces of pcodes that translated from a single instruction.
There are several reasons I need this, for example:

- The meaning of some instruction (such as pointer authentication instruction under arm arch) is already known when given constant register values. And I want to modify the pcodes so it reveals the semantic directly. Also, I've met a certain case of obfuscation that the semantic of that pointer authentication is modified. In that case, I want the semantic to be revealed even more eagerly.
- Just like your example, I want to also add the `try-catch` part when I'm analyzing the C++ exception-enabled program. Current analysis does not cope with C++ exception well especially in decompiled code. But since I'm not directly modifying Ghidra, I want this to be dynamic enough to be carried in a Python(Jython) script. In this case, I need the ability to modify/inject the pcode dynamically. I've checked your suggestion above, but I don't know how to make that dynamic without modifying the underling `x86_64.slaspec`, `x86.sinc`, etc. Can I manually construct a `SleighLanguage` to at least use my own modification without directly modifying released processor files? (Although this means I will have to modify each architecture's `slaspec` to provide my custom context register, and architectures like x86 might need to modify several files).

Another question might be, what can be unsafe/broken if it is possible to dynamically construct and set `PcodeInject`? Especially only to adding ones without modifying existing ones?

**Comment by Danil6969** [comment]
Solutions:
1) After analysis set some flag to 1 and provide constant value taking up 2 context registers for this, remember that largest size of context register is 4 bytes, use bit logic for 8 sized values and more registers
2) There's no need to construct objects for this particular task and not all targets support compilation with exceptions, remember this (as well as thread variables, rtti, other features). For example, if there's no delphi compiler for powerpc then why put "except" pcodeop here, in ppc language? Also you only need to modify ia.sinc as it is included in almost all x86. Anyway you will have to check other cases so it's made separately.

**Comment by Escapingbug** [comment]
> Solutions:
> 
> 1. After analysis set some flag to 1 and provide constant value taking up 2 context registers for this, remember that largest size of context register is 4 bytes, use bit logic for 8 sized values and more registers
> 2. There's no need to construct objects for this particular task and not all targets support compilation with exceptions, remember this (as well as thread variables, rtti, other features). For example, if there's no delphi compiler for powerpc then why put "except" pcodeop here, in ppc language? Also you only need to modify ia.sinc as it is included in almost all x86. Anyway you will have to check other cases so it's made separately.

I'm not sure if I've made my whole point clear. My point is, all these stuff are for **a single binary** (even for case 2, because I'm not yet sure my exception handling is correct, I'd like to script it and test on more binaries).
But modifying the architecture specification (by adding context register) does not seem to be dynamic enough for a single binary.
I don't think modifying the processor's specification only to analyze some specific binary is a good solution to this, unless it is possible to be dynamically introduced only for this binary (via manually construction of the `SleighLanguage` or so).
That's why I'm searching for dynamic solution, so I can dynamically adjust some pcode only for this binary.

**Comment by Escapingbug** [comment]
Any idea? Or can we add this?
